### PR TITLE
Simplify HivePageSource bucket filtering

### PR DESCRIPTION
**Pull request by pettyjamesm**
Extracted and adapted changes from https://github.com/prestodb/presto/pull/15373

Adds an early return path for when all rows or no rows are filtered as inelligible as part of bucket adaptation in `HivePageSource`

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/testing/MaterializedResult.java [hunk 1]
```diff
@@ -397,7 +397,7 @@ public static MaterializedResult materializeSourceDataStream(ConnectorSession se
         while (!pageSource.isFinished()) {
             Page outputPage = pageSource.getNextPage();
             if (outputPage == null) {
-                break;
+                continue;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/testing/MaterializedResult.java:400]: That's a bug fix. Please make it into a separate commit.
Is it possible to test? Are we having similar problem somewhere else?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java [hunk 2]
```diff
@@ -631,7 +626,14 @@ public IntArrayList computeEligibleRowIds(Page page)
                     ids.add(position);
                 }
             }
-            return ids;
+            int keptRowCount = ids.size();
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java:629]: kept -> retained

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java [hunk 3]
```diff
@@ -631,7 +626,14 @@ public IntArrayList computeEligibleRowIds(Page page)
...
+            if (keptRowCount == 0) {
+                return null; // Empty page after filtering
+            }
+            if (keptRowCount == page.getPositionCount()) {
+                return page; // Unchanged after filtering
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java:634]: i think the comments are redundant, but up to you

**Review by findepi — APPROVED** [review]

File: presto-main/src/main/java/io/prestosql/testing/MaterializedResult.java (hunk 1 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/testing/MaterializedResult.java:400]: Moved to separate commit. Unsure if the issue was a problem anywhere else before but I suspect not since tests passed before this change and still pass afterwards.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java (hunk 3 above)
Comment (pettyjamesm) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java]: Removed.

**Comment by findepi** [comment]
Merged, thanks!
